hi-ml-azure/src/health_azure/traverse.py

```diff
@@ -0,0 +1,234 @@
+#  ------------------------------------------------------------------------------------------
+#  Copyright (c) Microsoft Corporation. All rights reserved.
+#  Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+#  ------------------------------------------------------------------------------------------
+import enum
+import logging
+from io import StringIO
+from typing import Any, Dict, Union, List, Optional
+
+import param
+from ruamel import yaml
+
+
+def is_basic_field(o: Any) -> bool:
+    """
+    Returns True if the given object is an instance of a basic simple datatype: string, integer, float.
+    """
+    return isinstance(o, (str, int, float))
+
+
+def is_enum(o: Any) -> bool:
+    """
+    Returns True if the given object is a subclass of enum.Enum.
+
+    :param o: The object to inspect.
+    :return: True if the object is an enum, False otherwise.
+    """
+    return isinstance(o, enum.Enum)
+
+
+def get_all_writable_attributes(o: Any) -> Dict[str, Any]:
+    """
+    Returns all writable attributes of an object, by resorting to the "vars" method. For object that derive
+    from param.Parameterized, it returns all params that are not constant and not readonly.
+
+    :param o: The object to inspect.
+    :return: A dictionary mapping from attribute name to its value.
+    """
+    def _is_private(s: str) -> bool:
+        return s.startswith("_")
+
+    result = {}
+    if isinstance(o, param.Parameterized):
+        for param_name, p in o.params().items():
+            if _is_private(param_name):
+                logging.debug(f"get_all_writable_attributes: Skipping private field {param_name}")
+            elif p.constant:
+                logging.debug(f"get_all_writable_attributes: Skipping constant field {param_name}")
+            elif p.readonly:
+                logging.debug(f"get_all_writable_attributes: Skipping readonly field {param_name}")
+            else:
+                result[param_name] = getattr(o, param_name)
+        return result
+    try:
+        for name, value in vars(o).items():
+            if _is_private(name):
+                logging.debug(f"get_all_writable_attributes: Skipping private field {name}")
+            else:
+                result[name] = value
+        return result
+    except TypeError:
+        raise ValueError("This function cannot be used on objects that do not support the 'vars' operation")
+
+
+def _object_to_dict(o: Any) -> Union[Optional[str], Dict[str, Any]]:
```

Comment:
Does this function work if the object being passed in is a list?

Reply:
it does now, thanks for spotting!